### PR TITLE
feat: implement transaction [WPB-11400] [WPB-10917]

### PR DESCRIPTION
**Pull request by SimonThormeyer**
# What's new in this PR
Transaction implementation

----
##### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: crypto/src/lib.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@
 //! The goal is provide a easier and less verbose API to create, manage and interact with MLS
 //! groups.
 #![doc = include_str!("../../README.md")]
-#![deny(missing_docs)]
+#![cfg_attr(not(test), deny(missing_docs))]
```
Comment (istankovic) [inline crypto/src/lib.rs:23]: Why add this now?

**Comment by istankovic** [comment]
Do we have a test for `ContextState::Invalid` i.e. a test that causes the context to become invalid and then checks that we return proper errors when attempting to use the context further?

File: keystore/src/entities/platform/generic/proteus/identity.rs [hunk 2]
```diff
@@ -110,17 +111,15 @@ impl EntityBase for ProteusIdentity {
...
-        let transaction = conn.transaction()?;
-
+#[cfg_attr(target_family = "wasm", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait::async_trait)]
+impl EntityTransactionExt for ProteusIdentity {
+    async fn save(&self, transaction: &TransactionWrapper<'_>) -> crate::CryptoKeystoreResult<()> {
         use rusqlite::ToSql as _;
         transaction.execute(
-            "INSERT INTO proteus_identities (sk, pk) VALUES (?, ?)",
+            "INSERT INTO proteus_identities (sk, pk) SELECT ?, ? WHERE (SELECT COUNT(*) FROM proteus_identities) = 0",
```
Comment (istankovic) [inline keystore/src/entities/platform/generic/proteus/identity.rs:122]: What is the reason for this change?

Comment (SimonThormeyer) [reply keystore/src/entities/platform/generic/proteus/identity.rs:122]: To keep the logic that was present on current main, which would do an early return in this function:
```
if Self::count(conn).await? == 1 {
            return Ok(());
        }
```

File: keystore/src/entities/mod.rs [hunk 3]
```diff
@@ -261,6 +261,12 @@ cfg_if::cfg_if! {
 
         pub trait Entity: EntityBase {
             fn id_raw(&self) -> &[u8];
+            
+            /// The query results that are obtained during a transaction
+            /// from the transaction cache and the database are merged by this key.
```
Comment (istankovic) [inline keystore/src/entities/mod.rs:297]: Appreciate the comment!

File: crypto/src/lib.rs (hunk 1 above)
Comment (SimonThormeyer) [reply crypto/src/lib.rs]: No idea, removed it in [chore: fix docs](https://github.com/wireapp/core-crypto/pull/732/commits/fd93bce72ef8b8867e76b3f37466dc9cd8f30088).

File: keystore/src/entities/platform/generic/proteus/identity.rs (hunk 2 above)
Comment (SimonThormeyer) [reply keystore/src/entities/platform/generic/proteus/identity.rs:122]: Just realized that this won't work as expected if there is a record in the persisted DB but none in the cache.

File: keystore/src/transaction.rs [hunk 4]
```diff
@@ -80,379 +75,396 @@ impl EntityId {
...
-            MlsEntityId::EncryptionKeyPair => Self::EncryptionKeyPair(id.into()),
-            MlsEntityId::EpochEncryptionKeyPair => Self::EpochEncryptionKeyPair(id.into()),
-            MlsEntityId::GroupState => Self::PersistedMlsGroup(id.into()),
+            EntityId::ProteusIdentity(vec) => vec.as_slice().into(),
+            EntityId::ProteusSession(id) => id.as_slice().into(),
+            EntityId::ProteusPrekey(vec) => vec.as_slice().into(),
         }
     }
 
     fn from_collection_name(entity_id: &'static str, id: &[u8]) -> CryptoKeystoreResult<Self> {
```
Comment (istankovic) [inline keystore/src/transaction.rs]: I feel this would be a good candidate for `const fn`.

File: keystore/src/transaction.rs [hunk 5]
```diff
@@ -80,379 +75,396 @@ impl EntityId {
...
+        Ok(())
     }
-}
 
-impl KeystoreTransaction {
-    pub fn new(conn: Connection) -> Self {
-        Self {
-            conn,
-            operations: Default::default(),
+    pub(crate) async fn find<E>(&self, id: &[u8]) -> CryptoKeystoreResult<Option<Option<E>>>
```
Comment (istankovic) [inline keystore/src/transaction.rs]: Why do we have two `Option`s in a row in the result?

File: keystore/src/transaction.rs [hunk 6]
```diff
@@ -80,379 +75,396 @@ impl EntityId {
...
+            .merge_records(cached_records, persisted_records, EntityFindParams::default())
+            .await)
     }
 
-    pub async fn child_groups<
-        E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>
-            + crate::entities::PersistedMlsGroupExt
-            + Sync,
-    >(
+    async fn merge_records<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
```
Comment (istankovic) [inline keystore/src/transaction.rs]: It would be nice to add a comment describing how the merging is actually performed, i.e. how you get from input records to the output.

**Comment by istankovic** [comment]
We need to resolve code duplication in `src/transaction/dynamic_dispatch/{mls.rs,proteus_and_mls.rs}`.

File: keystore/src/connection/mod.rs [hunk 7]
```diff
@@ -192,8 +194,31 @@ impl Connection {
...
+    ) -> CryptoKeystoreResult<Vec<E>> {
+        let mut conn = self.conn.lock().await;
+        let persisted_records = entity.child_groups(conn.deref_mut()).await?;
+
+        let transaction_guard = self.transaction.lock().await;
+        let Some(transaction) = transaction_guard.as_ref() else {
+            return Ok(persisted_records);
+        };
+        transaction.child_groups(entity, persisted_records).await
     }
```
Comment (istankovic) [inline keystore/src/connection/mod.rs:225]: I don't understand this function, could you please described what it is supposed to do?

File: keystore/src/connection/mod.rs [hunk 8]
```diff
@@ -227,10 +277,21 @@ impl FetchFromDatabase for Connection {
...
+        };
+        transaction.find_many(persisted_records, ids).await
     }
 
     async fn count<E: Entity<ConnectionType = KeystoreDatabaseConnection>>(&self) -> CryptoKeystoreResult<usize> {
+        if self.transaction.lock().await.is_some() {
+            // Unfortunately, we have to do this because of possible record id overlap 
+            // between cache and db.
+            return Ok(self.find_all::<E>(Default::default()).await?.len());
+        };
```
Comment (istankovic) [inline keystore/src/connection/mod.rs:330]: This looks suspicious. Isn't `find_all` going to try to acquire the same lock that we got here?

File: keystore/src/entities/platform/generic/mls/credential.rs [hunk 9]
```diff
@@ -33,6 +33,13 @@ impl Entity for MlsCredential {
     fn id_raw(&self) -> &[u8] {
         self.id.as_slice()
     }
+
+    fn merge_key(&self) -> Vec<u8> {
+        let mut merge_key = self.id_raw().to_vec();
+        // Credentials are unique by id and type, the type is contained in these bytes.
+        merge_key.extend(self.credential.to_vec());
```
Comment (istankovic) [inline keystore/src/entities/platform/generic/mls/credential.rs:40]: Can we just iterate over `self.credential` in  order to avoid creating a new `Vec`, just so we can iterate over it and then immediately discard it?

File: crypto/src/mls/client/identities.rs [hunk 10]
```diff
@@ -5,106 +5,120 @@ use crate::{
...
-        self.0.get(&sc)?.iter().find(|c| {
+    ) -> Option<CredentialBundle> {
+        let credential_bundles_by_signature_schemes = self.0.lock().await;
+        let credential_bundles = credential_bundles_by_signature_schemes.get(&sc)?;
+        credential_bundles.iter().find(|c| {
             let ct_match = ct == c.credential.credential_type().into();
             let pk_match = c.signature_key.public() == pk.as_slice();
             ct_match && pk_match
-        })
+        }).cloned()
```
Comment (istankovic) [inline crypto/src/mls/client/identities.rs]: Why do we clone bundles now?

File: keystore/src/transaction.rs (hunk 5 above)
Comment (SimonThormeyer) [reply keystore/src/transaction.rs]: Thanks for pointing out that this may need clarification! I'm adding the following docstring
```rust
/// The result of this function will have different contents for different scenarios: 
/// * `Some(Some(E))` - the transaction cache contains the record
/// * `Some(None)` - the deletion of the record has been cached
/// * `None` - there is no information about the record in the cache 
```

File: keystore/src/transaction.rs (hunk 6 above)
Comment (SimonThormeyer) [reply keystore/src/transaction.rs]: Thanks for pointing this out.
Added this docstring

```rust
    /// Build a single list of unique records from two potentially overlapping lists.
    /// In case of overlap, records in `records_a` are prioritized.
    /// Identity from the perspective of this function is determined by the output of [crate::entities::Entity::merge_key].
    ///
    /// Further, the output list of records is built with respect to the provided [EntityFindParams]
    /// and the deleted records cached in this [Self] instance.     

```

File: keystore/src/entities/platform/generic/proteus/identity.rs (hunk 2 above)
Comment (SimonThormeyer) [reply keystore/src/entities/platform/generic/proteus/identity.rs]: Not an issue because the calling site takes care of the case that is handled here. Resolving.

File: keystore/src/connection/mod.rs (hunk 7 above)
Comment (SimonThormeyer) [reply keystore/src/connection/mod.rs:225]: It collects all records that have the  id of `entity` as `parent_id`. I don't know in what context this may be needed, AFAICT this function isn't used anywhere. Should we remove it in this PR or in another?

File: keystore/src/connection/mod.rs (hunk 8 above)
Comment (SimonThormeyer) [reply keystore/src/connection/mod.rs:330]: The tests run and execute this codepath without a deadlock. Calling `.is_some()` probably drops the lock.

File: crypto/src/mls/client/identities.rs (hunk 10 above)
Comment (SimonThormeyer) [reply crypto/src/mls/client/identities.rs]: Couldn't figure out a way to return references out of the mutex; Is there one?

File: keystore/src/entities/platform/generic/mls/credential.rs (hunk 9 above)
Comment (istankovic) [reply keystore/src/entities/platform/generic/mls/credential.rs]: Perfect!

**Comment by SimonThormeyer** [comment]
> Do we have a test for `ContextState::Invalid` i.e. a test that causes the context to become invalid and then checks that we return proper errors when attempting to use the context further?

Added a test in the typescript wrapper test.

**Comment by istankovic** [comment]
If an error occurs during `context.finish()`, we will propagate the error to the client, however we won't set the context's state to invalid. This opens up a possibility for the client to continue using the context (e.g. remembering the context variable in the scope outside transaction).

File: crypto-ffi/src/wasm/context/mod.rs [hunk 11]
```diff
@@ -18,6 +18,8 @@ use wasm_bindgen::prelude::wasm_bindgen;
 use wasm_bindgen::JsValue;
 use wasm_bindgen_futures::future_to_promise;
 
+mod e2ei;
```
Comment (istankovic) [inline crypto-ffi/src/wasm/context/mod.rs:21]: Why is the the one from `generic` `pub`, and this one not?

**Comment by SimonThormeyer** [comment]
> If an error occurs during `context.finish()`, we will propagate the error to the client, however we won't set the context's state to invalid. This opens up a possibility for the client to continue using the context (e.g. remembering the context variable in the scope outside transaction).

fixed in [chore: fix transactions in bindings](https://github.com/wireapp/core-crypto/pull/732/commits/307a045dfbd4ebd5a14b6b2d1f7cb04fc2638dea).

File: crypto-ffi/src/wasm/context/mod.rs (hunk 11 above)
Comment (SimonThormeyer) [reply crypto-ffi/src/wasm/context/mod.rs]: Good catch, made it consistent to be pub in both places.

File: interop/src/clients/corecrypto/ffi.rs [hunk 12]
```diff
@@ -101,85 +92,118 @@ impl<'a> EmulatedMlsClient for CoreCryptoFfiClient<'a> {
...
     async fn session_from_prekey(&mut self, session_id: &str, prekey: &[u8]) -> Result<()> {
-        let _ = self.cc.proteus_session_from_prekey(session_id, prekey)?;
+        let _ = self
+            .cc
+            .proteus_session_from_prekey(session_id.to_string(), prekey.to_vec())
+            .await?;
         Ok(())
     }
 
+    #[allow(deprecated)]
```
Comment (istankovic) [inline interop/src/clients/corecrypto/ffi.rs:182]: Why do we need this?

File: keystore/src/transaction/mod.rs [hunk 13]
```diff
@@ -283,30 +282,58 @@ impl KeystoreTransaction {
...
+                        (identifier_10, MlsPendingMessage),
+                        (identifier_11, E2eiEnrollment),
+                        (identifier_12, E2eiRefreshToken),
+                        (identifier_13, E2eiAcmeCA),
+                        (identifier_14, E2eiIntermediateCert),
+                        (identifier_15, E2eiCrl)
+                    ]
+                );
+            }
+        }
```
Comment (istankovic) [inline keystore/src/transaction/mod.rs:336]: Can we make this more concise by moving the Proteus-specific parts under their own `cfg_if` block?

File: interop/src/clients/corecrypto/ffi.rs (hunk 12 above)
Comment (SimonThormeyer) [reply interop/src/clients/corecrypto/ffi.rs:182]: WPB-12132

File: crypto-ffi/bindings/jvm/src/main/kotlin/com/wire/crypto/client/CoreCryptoCentral.kt [hunk 14]
```diff
@@ -34,11 +34,19 @@ private class Callbacks : CoreCryptoCallbacks {
...
+            try {
+                result = block(CoreCryptoContext(context))
+            } catch (e: Throwable) {
+                error = e
+            }
         }
     })
+    if (error != null) {
+        throw error as Throwable
+    }
```
Comment (istankovic) [inline crypto-ffi/bindings/jvm/src/main/kotlin/com/wire/crypto/client/CoreCryptoCentral.kt:49]: Do we have a test for this?

File: crypto-ffi/bindings/js/CoreCrypto.ts [hunk 15]
```diff
@@ -1043,15 +1043,27 @@ export class CoreCrypto {
...
+                        error = CoreCryptoError.fromStdError(
+                            e as Error
+                        ) as CoreCryptoError;
+                    }
+                }
+            },
+        });
+        if (error !== null) {
+            throw error;
+        }
```
Comment (istankovic) [inline crypto-ffi/bindings/js/CoreCrypto.ts:1066]: Do we have a test for this?

File: crypto-ffi/bindings/js/test/CoreCrypto.test.js [hunk 16]
```diff
@@ -225,6 +225,52 @@ test("can use groupInfo enums", async () => {
...
+                1
+            );
+        } catch (e) {
+            error = e;
+        }
+
+        return error;
+    });
+
+    expect(error.rustStackTrace).toBe("CryptoError(InvalidContext)");
```
Comment (istankovic) [inline crypto-ffi/bindings/js/test/CoreCrypto.test.js:268]: Thank you for this test! :purple_heart:

File: crypto-ffi/bindings/jvm/src/main/kotlin/com/wire/crypto/client/CoreCryptoCentral.kt (hunk 14 above)
Comment (SimonThormeyer) [reply crypto-ffi/bindings/jvm/src/main/kotlin/com/wire/crypto/client/CoreCryptoCentral.kt:55]: Test was added

File: crypto-ffi/bindings/js/CoreCrypto.ts (hunk 15 above)
Comment (SimonThormeyer) [reply crypto-ffi/bindings/js/CoreCrypto.ts:1072]: Test was added

File: crypto-ffi/bindings/js/test/CoreCrypto.test.js [hunk 17]
```diff
@@ -315,6 +315,58 @@ test("JS Error is propagated by transaction", async () => {
     await page.close();
     await ctx.close();
 });
+test("Transaction is rolled back after error", async () => {
```
Comment (istankovic) [inline crypto-ffi/bindings/js/test/CoreCrypto.test.js:318]: Nit: missing one empty line between tests.

**Review by istankovic — APPROVED** [review]
Massive work, great job. :partying_face:
